### PR TITLE
Fix #1691, Add External Time Source Functional Tests

### LINKED ISSUE #1691: Add functional tests for cFE External Time Source APIs
**Is your feature request related to a problem? Please describe.**
Need open source functional tests for certifiability

**Describe the solution you'd like**
Add functional tests for cFE External Time Source APIs
    CFE_TIME_ExternalTone - Provides the 1 Hz signal from an external source.
    CFE_TIME_ExternalMET - Provides the Mission Elapsed Time from an external source.
    CFE_TIME_ExternalGPS - Provide the time from an external source that has data common to GPS receivers.
    CFE_TIME_ExternalTime - Provide the time from an external source that measures time relative to a known epoch.
    CFE_TIME_RegisterSynchCallback - Registers a callback function that is called whenever time synchronization occurs.
    CFE_TIME_UnregisterSynchCallback - Unregisters a callback function that is called whenever time synchronization occurs.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Requester Info**
Alex Campbell GSFC

### PR DESCRIPTION
**Pull request by zanzaben**
**Describe the contribution**
Fixes #1691
Add External Time Source Functional Tests

**Testing performed**
Build and run unit test

**Expected behavior changes**
No impact to behavior

**System(s) tested on**
Ubuntu 20.04

**Contributor Info - All information REQUIRED for consideration of pull request**
Alex Campbell GSFC

### CODE REVIEW

File: modules/cfe_testcase/src/time_external_test.c [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+#endif
+}
+
+void TestCallback(void)
+{
+    UtPrintf("Testing: CFE_TIME_RegisterSynchCallback, CFE_TIME_UnregisterSynchCallback");
+
+    UtAssert_INT32_EQ(CFE_TIME_RegisterSynchCallback(&TestCallbackFunction), CFE_SUCCESS);
+    UtAssert_INT32_EQ(CFE_TIME_RegisterSynchCallback(&TestCallbackFunction2), CFE_TIME_TOO_MANY_SYNCH_CALLBACKS);
+
```
Comment (skliper) [inline modules/cfe_testcase/src/time_external_test.c:76]: How about adding a 2 sec delay and check the CallbackFunction is called at least once?  Could use the FT count global...  Not a big deal but would be an easy functional check.

File: modules/cfe_testcase/src/time_external_test.c [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+void TestExternal(void)
+{
+#if ((CFE_PLATFORM_TIME_CFG_SRC_MET == true) || (CFE_PLATFORM_TIME_CFG_SRC_GPS == true) || \
+     (CFE_PLATFORM_TIME_CFG_SRC_TIME == true))
+    CFE_TIME_SysTime_t time = {1000, 0};
+#endif
+
+    UtPrintf("Testing: CFE_TIME_ExternalTone, CFE_TIME_ExternalMET, CFE_TIME_ExternalGPS, CFE_TIME_ExternalTime");
+
```
Comment (skliper) [inline modules/cfe_testcase/src/time_external_test.c:54]: Worth a UtPrintf or comment here that the following time calls could impact the system timekeeping.  Likely impact is incorrect time for one update cycle, a rejected external time update, multiple tone's or external updates detected, or similar.  The call here confirms the implementation is properly included, more extensive timekeeping verification is a project/system responsibility.

**Review by skliper — CHANGES_REQUESTED** [review]

**Comment by astrogeco** [comment]
[CCB:2021-08-11](https://github.com/nasa/cFS/wiki/Community-CCB-Agenda:-August-11-2021) Needs more work

**Comment by zanzaben** [comment]
This current implementation of verifying the callback function is not 100% reliable yet.

File: modules/cfe_testcase/src/time_external_test.c [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+void TestCallback(void)
+{
+    CFE_FT_Global.Count = 1;
+
+    UtPrintf("Testing: CFE_TIME_RegisterSynchCallback, CFE_TIME_UnregisterSynchCallback");
+
+    UtAssert_INT32_EQ(CFE_TIME_RegisterSynchCallback(&TestCallbackFunction), CFE_SUCCESS);
+    UtAssert_INT32_EQ(CFE_TIME_RegisterSynchCallback(&TestCallbackFunction2), CFE_TIME_TOO_MANY_SYNCH_CALLBACKS);
+
+    OS_TaskDelay(2000);
```
Comment (skliper) [inline modules/cfe_testcase/src/time_external_test.c:57]: How about 2500 to give a little margin?

File: modules/cfe_testcase/src/time_external_test.c [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    if (CFE_FT_Global.Count < 2)
+    {
+        UtAssert_MIR("Timing performance of this system is insufficient to test CFE_TIME_RegisterSynchCallback");
+    }
+
+    CFE_FT_Global.Count = 1;
+    UtAssert_INT32_EQ(CFE_TIME_UnregisterSynchCallback(&TestCallbackFunction), CFE_SUCCESS);
+    UtAssert_INT32_EQ(CFE_TIME_UnregisterSynchCallback(&TestCallbackFunction), CFE_TIME_CALLBACK_NOT_REGISTERED);
+
+    OS_TaskDelay(2000);
```
Comment (skliper) [inline modules/cfe_testcase/src/time_external_test.c:67]: same 2500

File: modules/cfe_testcase/src/time_external_test.c [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    UtPrintf("Testing: CFE_TIME_RegisterSynchCallback, CFE_TIME_UnregisterSynchCallback");
+
+    UtAssert_INT32_EQ(CFE_TIME_RegisterSynchCallback(&TestCallbackFunction), CFE_SUCCESS);
+    UtAssert_INT32_EQ(CFE_TIME_RegisterSynchCallback(&TestCallbackFunction2), CFE_TIME_TOO_MANY_SYNCH_CALLBACKS);
+
+    OS_TaskDelay(2000);
+    if (CFE_FT_Global.Count < 2)
+    {
+        UtAssert_MIR("Timing performance of this system is insufficient to test CFE_TIME_RegisterSynchCallback");
```
Comment (skliper) [inline modules/cfe_testcase/src/time_external_test.c:60]: How about "...likely insufficient..."  Comment might help also, something about the callback only gets called if the elapsed time between tones is within the configured tolerance.  Manual inspection is required to determine if there is an actual failure in registering or executing the callback or if it's a timing issue within the system under test, so the test is not automatically failed here but requires further analysis.

**Review by skliper — CHANGES_REQUESTED** [review]
Embedded...

**Review by skliper — APPROVED** [review]

**Pull request merged**